hail/src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -416,38 +416,47 @@ object Simplify {
     case InsertFields(SelectFields(old, _), Seq(), Some(insertFieldOrder)) =>
       SelectFields(old, insertFieldOrder)
 
-    case Let(Seq(), body) =>
+    case Block(Seq(), body) =>
       body
 
-    case Let(xs, Let(ys, body)) =>
-      Let(xs ++ ys, body)
+    case Block(xs, Block(ys, body)) =>
+      Block(xs ++ ys, body)
 
     // assumes `NormalizeNames` has been run before this.
-    case Let(Let.Nested(before, after), body) =>
-      def numBindings(b: (String, IR)): Int =
-        b._2 match {
-          case let: Let => 1 + let.bindings.length
+    case Block(bindings, body) if bindings.exists(_.value.isInstanceOf[Block]) =>
+      def numBindings(b: Binding): Int =
+        b.value match {
+          case let: Block => 1 + let.bindings.length
           case _ => 1
         }
 
       val newBindings =
-        new BoxedArrayBuilder[(String, IR)](
-          after.foldLeft(before.length)((sum, binding) => sum + numBindings(binding))
+        new BoxedArrayBuilder[Binding](
+          bindings.foldLeft(0)((sum, binding) => sum + numBindings(binding))
         )
 
-      newBindings ++= before
-
-      after.foreach {
-        case (name: String, ir: Let) =>
-          newBindings ++= ir.bindings
-          newBindings += name -> ir.body
-        case (name, value) =>
-          newBindings += name -> value
+      bindings.foreach {
+        case Binding(name, ir: Block, scope) =>
+          newBindings ++= (if (scope == Scope.EVAL) ir.bindings
+                           else ir.bindings.map {
+                             case Binding(name, value, Scope.EVAL) => Binding(name, value, scope)
```

Comment:
I'd prefer if this rule was simpler and didn't deal with scope changes.

Reply:
Are you requesting a change? This rule is at the core of this PR: enforcing a stronger normal form on the IR, where arbitrary nestings of alternating `Let` and `AggLet` are flattened to a single level. That adds to the local complexity of this rule, because our binding structure is complicated. But until we can make a larger redesign to simplify the binding structure, this complexity is necessary to enforce the normal form, which in turn makes the IR as a whole simpler (anything consuming IR can make stronger assumptions about the form of the IR). In particular, this normal form is needed for the SSA printer, where the new `Block` node exactly corresponds to a block in the SSA form.

Reply:
OK. Is the partial function safe?

Reply:
Yes, as discussed [here](https://github.com/hail-is/hail/pull/14402/files#r1521650912) (I know that was a while ago). This is in the `scope != EVAL` branch, and our scoping rules prevent references to aggregation (including defining agg bindings) inside the value of an agg binding. (I'm actively working on making all this much simpler.)